### DISCUSSION TITLE
New feature - redaction based on values in the data

### DISCUSSION THREAD

**Discussion by elsalant** [fybrik/fybrik#529] [Ideas]
Support for redaction of data based on a key value returned in the data.  For example, consider a data source that describes videos in a repository.  Each video is described by a record, which has a URL pointing to the video, and a tag describing what organization owns the video.  We need to be able to support a rule which states that the caller can only see a video URL if the caller's (requestor) organization is the same as the owning organization in the video record.

From a discussion with Roee, this could be supported as follows:
1. The data resource (asset) yaml would specify in addition to the PII fields, the fields of interest for policy logic.  (E.g, "organization" in the above example).  This specification needs to be able to support a nested structure (e.g. nested JSON)
2. The OPA policy needs to return the redaction action that needs to be taken, for example as an SQL statement
3. The data module calling OPA needs to be able to execute the returned SQL statement on retrieved data to achieve redaction.

Additionally, the nesting hierarchy should only be specified in the asset yaml, but not in the policy definition (rego).  For example, in the asset file, we would specify the full path (e.g. "video.creator.org"), and assign it a tag ("owning_org").  The policy would only use the tag, e.g. "owning_org == requesting_org"

**Comment by simanadler** [comment]
@elsalant This is what I often call row level (or object level) enforcement.  There are several areas that need to be addressed to provide this functionality.
1) Policy Manager - Needs to know how to describe row/object level policies
2) Policy Manager - Needs to know how to describe row/object level policy decisions
2) Enforcement Point (M4D Modules) need(s) to know how to enforce row level / object level policy decisions

In the M4D architecture it is the M4DApplication controller that confers with the Policy Manager to receive the policy decisions.  It then passes those decisions to the relevant module.  The module doesn't call OPA as you mentioned in #3.

**Reply by elsalant** [reply]
I think that having the Policy Manager return an SQL statement for the Modules to execute would address these issues - assuming that the PEP (Module) embeds some type of SQL engine.  
I think it there are still some other issues to think about - are there cases where the Policy Manager just returns a list of data fields to redact and other cases where the Policy Mgr returns SQL to be executed?

**Reply by simanadler** [reply]
@elsalant  Policy managers do not return sql.  Their policy decision point returns the decision (allow, deny) and if allow, possible transforms that should be performed on particular components of the data set.  A separate component known as the policy enforcement point (PEP), receives the policy decision and does the enforcement in the manner relevant to the particular storage type.  I.e. The PEP would translate it into sql.

**Reply by elsalant** [reply]
There is no reason that the Rego policies cannot return an SQL string.  In fact, it makes more sense, IMO, to put the SQL inside the policies than to attempt to have the PEP translate it to SQL.  In this manner, the PEP (module) remains agnostic to the data schema.

**Reply by simanadler** [reply]
@elsalant Architecturally I do not think this is the best approach.  You are moving the knowledge of the data model to the policy engine instead of the PEP.  It's supposed to be the opposite.  The beauty of M4D though is that it allows you to define the policies and the PEP (M4D modules) as you wish, without changing M4D.  :)

**Reply by elsalant** [reply]
@simanadler As I see things, the module needs both to be responsible for fetching data from the data source and acting as a PEP. I should only have one module to handle Arrow Flight format, or REST format.  The policy is very much related to the data  as it is the logical place to express schema information without having to change the module.  Let's look at a concrete example:  in one use case, a requester should only be allowed access to data records if the requester's organization (passed as input to the Policy Manager) is the same as the organization value contained inside of the data record.  The module has no idea where to find the organization value inside of the data record, since it has no knowledge of the schema of the data record.  In fact, different organizations may want to use the same REST module, but each organization has a different schema for their data records.  In this case, the knowledge of where to find the organization value inside of the record can be expressed in an SQL string in the Rego policy, and passed to the module.  The module simply executes the passed string and carries out the redaction based on the results - pretty much exactly as if were passed a list of columns to redact.

**Reply by roee88** [reply]
The knowledge of where to find the organization value inside of the record is sourced in the data catalog. The policy manager uses that knowledge to return appropriate action. The action described in this thread is a filter predicate.

> assuming that the PEP (Module) embeds some type of SQL engine

This is an assumption that you can omit if the action discussed here is not SQL but rather a filter predicate. This is a design choice to make and the decision should be made based on the applicability of the action type to different data models.

**Reply by elsalant** [reply]
@roee88 

> This is an assumption that you can omit if the action discussed here is not SQL but rather a filter predicate. This is a design choice to make and the decision should be made based on the applicability of the action type to different data models.

Who makes the design choice?  Is this the individual module or do we need to use the same language for the filter predicate for all modules?
